src/dcs/colmod.rs

```diff
@@ -0,0 +1,79 @@
+//! Module for the COLMOD instruction constructors
+
+use embedded_graphics_core::prelude::{PixelColor, RawData};
+
+use crate::{instruction::Instruction, Error};
+
+use super::DcsCommand;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Colmod(u8);
+
+impl Colmod {
+    ///
+    /// Set the color format that the MCU will be sending to this
+    /// display. Uses RawData::BITS_PER_PIXEL to decide.
+    ///
+    pub const fn new<PF>() -> Self
+    where
+        PF: PixelColor,
+    {
+        match PF::Raw::BITS_PER_PIXEL {
+            16 => Self(0b0101_0101),
+            18 => Self(0b0110_0110),
+            24 => Self(0b0110_0111), // not sure if this is right
+            _ => panic!("Incompatible PixelFormat size"),
+        }
+    }
```

Comment:
Yes you're right. I wonder how we can best handle this. It seems to me it should be a compile time "choice". This presents additional challenge since technically we should set the correct bits based on both the "bits per pixel" and the "which interface are you using" questions.

Reply:
I would suggest to only care about a compile time choice for the DBI color format at the moment. This is the format that is associated with the `display-interface` and is the one that used by the vast majority of users.

For the RGB/DPI format a simple enum should be enough to get started. The display driver crate is only one part of getting an RGB display to work. The other is the peripheral on the controller side, which generates the parallel RGB signals. In the future both sides should share some configuration to make this easier to setup and less error prone, but that is outside the scope of this crate.

Reply:
I've pushed the enum version in, but I'm not super happy about the ergonomics.

Reply:
I don't think it's too bad and is also something most users won't need to touch. What aspect of the ergonomics do you not like? That you need to manually specify the BPP for the color type?

Reply:
I was hoping there could be a way to unify the colors and di setups so that the pixel format and color model variants could be generic. E.g. instead of having to do a `ILI9486` for `Rgb565` and `Rgb666` that it could just be a ILI9486 for `PF: RgbColor` and the draw_pixels and pixel format selections could be generically handled since "technically" we know all the answers compile time.

Of course this was true for the colors only before (e.g. the `RawU18`). At this point neither the DI nor the `e-g::RgbColor` have trait level awareness of their bits per pixel.

This means that each model that has multiple color supports now need to be somewhat duplicated with `init_common` like setups. I just feel there's a way to remove that as well as the runtime setting for pixel format in general.

Probably work for later iterations though.

Reply:
Sounds like a good idea.

> Of course this was true for the colors only before (e.g. the `RawU18`). At this point neither the DI nor the `e-g::RgbColor` have trait level awareness of their bits per pixel.

Adding a bits per pixel constant to `RgbColor` might be a good idea for future e-g versions. But you can already calculate the bit depth from the `MAX_...` constants: `C::MAX_R.next_power_of_two() + C::MAX_G.next_power_of_two() + C::MAX_B.next_power_of_two()` (untested, but I think it should work and is also `const`)

Another option would be to add a new trait to add the necessary information:
```rust
trait DisplayColor {
    const BPP: BitsPerPixel;
}

impl DisplayColor for Rgb565 {
    const BPP: BitsPerPixel = BitsPerPixel::Sixteen;
}
```

Reply:
> Adding a bits per pixel constant to RgbColor might be a good idea for future e-g versions. But you can already calculate the bit depth from the MAX_... constants: C::MAX_R.next_power_of_two() + C::MAX_G.next_power_of_two() + C::MAX_B.next_power_of_two() (untested, but I think it should work and is also const)

This gives 24 with `Rgb666`, so manual trait is the only way for now

Reply:
Sorry I did get the functions mixed up, this should work: `C::MAX_R.trailing_ones() + C::MAX_G.trailing_ones() + C::MAX_B.trailing_ones()`

Reply:
Ah right. That works. I added it but I still think we can do better. If we expanded `WriteOnlyCommand` with something like `const BitsPerPixel: u8` `SetPixelFormat` could then be purely compile time, "auto-derived" from the `DI` and the `PixelColor` types.

Reply:
It would be nice to derive as much of the configuration at compile time. But there are a lot of possible configurations, which will probably make this hard. And IMO also very much outside the scope of this PR :wink: .

I think that `BitsPerPixel` is the wrong name for such a constant. The display interface only handles data and the pixel color to data translation should be part of the driver crate. There are so many different ways of packing pixel colors into data that this should remain the responsibility of the driver crate. But perhaps that is what you meant and you only chose an inaccurate name.

And then there are "fun" special cases like the ILI9486, which supports RGB565 via parallel interface, but not SPI. It would be nice to be able to check this a compile time, but it wouldn't be possible with the generic `display-interface`.

Reply:
> I think that `BitsPerPixel` is the wrong name for such a constant. The display interface only handles data and the pixel color to data translation should be part of the driver crate. There are so many different ways of packing pixel colors into data that this should remain the responsibility of the driver crate. But perhaps that is what you meant and you only chose an inaccurate name.

What would you name it? Also do you think the DBI and DPI should have different types then?

Reply:
> What would you name it?

I'm not very familiar with the `display-interface` crate and would need to look into this some more. But from what I can remember I didn't like the implementation very much the last time I looked at it. The crate also lacks support for less common parallel bus width, like 9, 18 and 24 bit.

IMO it's important to distinguish between color BPP and the interface bit width. A 16 BPP display could, for example, be connected via a 8 bit parallel bus or 16 bit parallel bus.

I think the display interface needs to provide more information in general to support all use cases. If you, for example, want to send RGB444 data to a ST7789 display the data must be packed differently for a 8 bit or 16 bit parallel interface:
![grafik](https://user-images.githubusercontent.com/226123/219757878-7fe7c6c9-7130-4482-baf4-ee370b26c7f4.png)
![grafik](https://user-images.githubusercontent.com/226123/219758025-ba88ace1-cbc9-4c7f-bf7c-1619e9db4582.png)
This can only be handled by the display driver, which needs to know the interface type and width to configure the display appropriately.

> Also do you think the DBI and DPI should have different types then?

The RGB interface will need its own set of types to share the configuration with the peripheral in the microcontroller that generates the signal, e.g. the LTDC in a STM32. This connection wouldn't be handled by `display-interface`.

Reply:
My main steps after this PR are as follows:
1. Separate `dcs` module into its own crate
2. Implement `ReadWriteCommand` in `display-interface`
3. Refactor `display-interface` to support these settings better

I'll probably just merge this PR as is now and then push for ergonomics on `SetPixelFormat` later. The main remaining decision is whether `BitsPerPixel` should be renamed and split into two types (e.g. one for DBI and one for DPI)

Reply:
> My main steps after this PR are as follows:
> 
> 1. Separate `dcs` module into its own crate

Do you have a use case in mind which requires that the DCS module is a separate crate? It would be a cleaner separation of concerns, but if no other driver uses the crate it would be additional maintenance effort with little benefit. If you don't mind the additional effort go ahead, but otherwise I would suggest to wait until this becomes necessary.

> I'll probably just merge this PR as is now and then push for ergonomics on `SetPixelFormat` later. The main remaining decision is whether `BitsPerPixel` should be renamed and split into two types (e.g. one for DBI and one for DPI)

IMO it's fine to use `BitsPerPixel` per pixel for both the DBI and DPI settings. They should be handled differently in other parts of the code, but `SetPixelFormat` is so low level that using the same type for both is the right thing to do.